### PR TITLE
PS3 Handler IISWebAppDeploy

### PR DESCRIPTION
**Pull request by 20shivangi**

### CODE REVIEW

File: Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/MsDeployOnTargetMachines.ps1 [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+function Get-MsDeployLocation
+{
+    param(
+    [Parameter(Mandatory=$true)]
+    [string]$regKeyPath
+    )
+    
+    if( -not (Test-Path -Path $regKeyPath))
+    {
+        throw (Get-VstsLocString -Key "MsDeployNotFoundError")
```
Comment (Ajay-MS) [inline Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/MsDeployOnTargetMachines.ps1]: This script will be executed on the remote machine. `VSTS task sdk` won't be available there. Hence `Get-VstsLocString` won't work.

Comment (20shivangi) [reply Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/MsDeployOnTargetMachines.ps1]: It is done

File: Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/DeployIISWebApp.ps1 [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        [string]$scriptToRun,
+        [string]$adminUserName,
+        [string]$adminPassword,
+        [string]$winrmProtocol,
+        [string]$testCertificate,
+        [string]$deployInParallel,
+        [string]$webDeployPackage
+    )
+
+    Write-Host "Starting deployment of IIS Web Deploy Package : $webDeployPackage"
```
Comment (Ajay-MS) [inline Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/DeployIISWebApp.ps1]: We can localize this string.

File: Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/DeployIISWebApp.ps1 [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+    $errorMessage = Invoke-RemoteDeployment -machinesList $machinesList -scriptToRun $scriptToRun -adminUserName $adminUserName -adminPassword $adminPassword -protocol $winrmProtocol -testCertificate $testCertificate -deployInParallel $deployInParallel
+
+    if(-not [string]::IsNullOrEmpty($errorMessage))
+    {
+        $helpMessage = "For more info please refer to http://aka.ms/iisextnreadme"
```
Comment (Ajay-MS) [inline Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/DeployIISWebApp.ps1]: Should help the message be localized.

File: Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/DeployIISWebApp.ps1 (hunk 2 above)
Comment (Ajay-MS) [reply Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/DeployIISWebApp.ps1]: Consider for all Write-host, write-output

File: Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/Main.ps1 [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+$adminPassword = Get-VstsInput -Name AdminPassword -Require
+$winrmProtocol = Get-VstsInput -Name WinrmProtocol -Require
+$testCertificate = Get-VstsInput -Name TestCertificate
+$webDeployPackage = Get-VstsInput -Name WebDeployPackage -Require
+$webDeployParamFile = Get-VstsInput -Name WebDeployParamFile
+$overRideParams = Get-VstsInput -Name OverRideParams
+$websiteName = Get-VstsInput -Name WebsiteName -Require
+$removeAdditionalFiles = Get-VstsInput -Name RemoveAdditionalFiles
+$excludeFilesFromAppData = Get-VstsInput -Name ExcludeFilesFromAppData
+$takeAppOffline = Get-VstsInput -Name TakeAppOffline
```
Comment (Ajay-MS) [inline Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/Main.ps1]: You can pass the argument ` -AsBool` which ensures only boolean input is provided.

File: Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/externals.json [hunk 5]
```diff
@@ -0,0 +1,21 @@
+
+{
+    "nugetv2": {
+        "VstsTaskSdk": {
+            "version": "0.7.1",
```
Comment (Ajay-MS) [inline Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/externals.json]: `0.7.1` is quite old version. Do we need to consider the latest version?

File: Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/DeployIISWebApp.ps1 (hunk 2 above)
Comment (20shivangi) [reply Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/DeployIISWebApp.ps1]: It is done

File: Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/Main.ps1 (hunk 4 above)
Comment (20shivangi) [reply Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/Main.ps1]: It is done

File: Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/externals.json (hunk 5 above)
Comment (20shivangi) [reply Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/externals.json]: Latest versions are giving some errors. Have tried all versions: 0.8.2 , 0.10.0 and 0.11.0

File: Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/DeployIISWebApp.ps1 (hunk 3 above)
Comment (20shivangi) [reply Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/DeployIISWebApp.ps1]: It is done.

**Comment by niadak** [comment]
Do we need to have package.json, task.lock.json files as well?

File: Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/task.json [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+      "helpMarkDown": "Select the option to deploy the website in-parallel on the target machines."
+    }
+  ],
+  "instanceNameFormat": "Deploy IIS App: $(WebDeployPackage)",
+  "execution": {
+    "PowerShell3": {
+      "target": "Main.ps1"
+    }
+  },
+  "messages": {
```
Comment (niadak) [inline Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/task.json]: Why we are adding messages here?

File: Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/Main.ps1 [hunk 7]
```diff
@@ -0,0 +1,50 @@
+[CmdletBinding()]
+param()
+
+$env:CURRENT_TASK_ROOTDIR = Split-Path -Parent $MyInvocation.MyCommand.Path
+
+Import-Module $env:CURRENT_TASK_ROOTDIR\ps_modules\VstsTaskSdk
+
+Trace-VstsEnteringInvocation $MyInvocation
```
Comment (niadak) [inline Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/Main.ps1]: This should be the first line

File: Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/Main.ps1 [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+# Import the loc strings.
+Import-VstsLocStrings -LiteralPath $env:CURRENT_TASK_ROOTDIR\task.json
```
Comment (niadak) [inline Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/Main.ps1]: Are we doing localization?

**Review by niadak — CHANGES_REQUESTED** [review]

File: Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/task.json (hunk 6 above)
Comment (20shivangi) [reply Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/task.json]: It is removed

File: Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/Main.ps1 (hunk 7 above)
Comment (20shivangi) [reply Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/Main.ps1]: It is done.

File: Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/Main.ps1 (hunk 8 above)
Comment (20shivangi) [reply Extensions/IISWebAppDeploy/Src/Tasks/IISWebAppDeploy/IISWebAppDeployV2/Main.ps1]: It is removed

**Review by niadak — APPROVED** [review]

**Pull request merged**